compiler/hash-typecheck/src/traits.rs

```diff
@@ -0,0 +1,128 @@
+use std::ops::ControlFlow;
+
+use crate::{env::HasTcEnv, errors::TcResult, options::normalisation::NormaliseResult};
+
+pub trait Operations<X>: HasTcEnv {
+    type TyNode;
+    type Node;
+
+    fn check(&self, item: &mut X, item_ty: Self::TyNode, item_node: Self::Node) -> TcResult<()>;
+
+    fn try_normalise(
+        &self,
+        item: X,
+        item_node: Self::Node,
+    ) -> NormaliseResult<ControlFlow<Self::Node>>;
+
+    fn unify(
+        &self,
+        src: &mut X,
+        target: &mut X,
+        src_node: Self::Node,
+        target_node: Self::Node,
+    ) -> TcResult<()>;
+}
+
+pub trait OperationsOnNode<X: Copy>: HasTcEnv {
+    type TyNode;
+
+    fn check_node(&self, item: X, item_ty: Self::TyNode) -> TcResult<()>;
+
+    fn try_normalise_node(&self, item: X) -> NormaliseResult<ControlFlow<X>>;
+
+    fn unify_nodes(&self, src: X, target: X) -> TcResult<()>;
+}
+
+impl<X: Copy, T: HasTcEnv + OperationsOnNode<X>> Operations<X> for T {
+    type TyNode = T::TyNode;
+    type Node = X;
+
+    fn check(&self, item: &mut X, item_ty: Self::TyNode, _: Self::Node) -> TcResult<()> {
+        self.check_node(*item, item_ty)
+    }
+
+    fn try_normalise(&self, item: X, _: Self::Node) -> NormaliseResult<ControlFlow<X>> {
+        self.try_normalise_node(item)
+    }
+
+    fn unify(&self, src: &mut X, target: &mut X, _: Self::Node, _: Self::Node) -> TcResult<()> {
+        self.unify_nodes(*src, *target)
+    }
+}
+
+pub trait RecursiveOperations<X>: HasTcEnv {
```

Comment:
whats the need and difference between `OperationsOn` and `RecursiveOperationsOn`?

Reply:
Simply put, recursive operations takes a closure. It is useful for things that add some "scoped" context like parameters, so that the closure is run while the parameters are still in scope. Again this is something I should document.